// @better-auth/expo/dist/client.js
import { t as PACKAGE_VERSION } from "./version-BXGK4jTV.js";
import { safeJSONParse } from "@better-auth/core/utils/json";
import { SECURE_COOKIE_PREFIX, parseSetCookieHeader, parseSetCookieHeader as parseSetCookieHeader$1, stripSecureCookiePrefix } from "better-auth/cookies";
import Constants from "expo-constants";
import * as Linking from "expo-linking";
import { AppState, Platform } from "react-native";
import { kFocusManager, kOnlineManager } from "better-auth/client";
//#region \0rolldown/runtime.js
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, { get: (a, b) => (typeof require !== "undefined" ? require : a)[b] }) : x)(function(x) {
	if (typeof require !== "undefined") return require.apply(this, arguments);
	throw Error("Calling `require` for \"" + x + "\" in an environment that doesn't expose the `require` function. See https://rolldown.rs/in-depth/bundling-cjs#require-external-modules for more details.");
});
//#endregion
//#region src/focus-manager.ts
var ExpoFocusManager = class {
	listeners = /* @__PURE__ */ new Set();
	subscription;
	isFocused;
	subscribe(listener) {
		this.listeners.add(listener);
		return () => {
			this.listeners.delete(listener);
		};
	}
	setFocused(focused) {
		if (this.isFocused === focused) return;
		this.isFocused = focused;
		this.listeners.forEach((listener) => listener(focused));
	}
	setup() {
		this.subscription = AppState.addEventListener("change", (state) => {
			this.setFocused(state === "active");
		});
		return () => {
			this.subscription?.remove();
		};
	}
};
function setupExpoFocusManager() {
	if (!globalThis[kFocusManager]) globalThis[kFocusManager] = new ExpoFocusManager();
	return globalThis[kFocusManager];
}
//#endregion
//#region src/online-manager.ts
var ExpoOnlineManager = class {
	listeners = /* @__PURE__ */ new Set();
	isOnline = true;
	unsubscribe;
	subscribe(listener) {
		this.listeners.add(listener);
		return () => {
			this.listeners.delete(listener);
		};
	}
	setOnline(online) {
		if (this.isOnline === online) return;
		this.isOnline = online;
		this.listeners.forEach((listener) => listener(online));
	}
	setup() {
		import("expo-network").then(({ addNetworkStateListener }) => {
			const subscription = addNetworkStateListener((state) => {
				this.setOnline(!!state.isInternetReachable);
			});
			this.unsubscribe = () => subscription.remove();
		}).catch(() => {
			this.setOnline(true);
		});
		return () => {
			this.unsubscribe?.();
		};
	}
};
function setupExpoOnlineManager() {
	if (!globalThis[kOnlineManager]) globalThis[kOnlineManager] = new ExpoOnlineManager();
	return globalThis[kOnlineManager];
}
//#endregion
//#region src/client.ts
if (Platform.OS !== "web") {
	setupExpoFocusManager();
	setupExpoOnlineManager();
}
function getSetCookie(header, prevCookie) {
	const parsed = parseSetCookieHeader$1(header);
	const toSetCookie = safeJSONParse(prevCookie) ?? {};
	parsed.forEach((cookie, key) => {
		const expiresAt = cookie["expires"];
		const maxAge = cookie["max-age"];
		if (maxAge !== void 0 && Number(maxAge) <= 0) {
			delete toSetCookie[key];
			return;
		}
		const expires = maxAge ? new Date(Date.now() + Number(maxAge) * 1e3) : expiresAt ? new Date(String(expiresAt)) : null;
		if (expires && expires.getTime() <= Date.now()) {
			delete toSetCookie[key];
			return;
		}
		toSetCookie[key] = {
			value: cookie["value"],
			expires: expires ? expires.toISOString() : null
		};
	});
	return JSON.stringify(toSetCookie);
}
function getCookie(cookie) {
	let parsed = {};
	try {
		parsed = JSON.parse(cookie);
	} catch {}
	return Object.entries(parsed).reduce((acc, [key, value]) => {
		if (value.expires && new Date(value.expires) < /* @__PURE__ */ new Date()) return acc;
		return acc ? `${acc}; ${key}=${value.value}` : `${key}=${value.value}`;
	}, "");
}
function getOAuthStateValue(cookieJson, cookiePrefix) {
	if (!cookieJson) return null;
	const parsed = safeJSONParse(cookieJson);
	if (!parsed) return null;
	const prefixes = Array.isArray(cookiePrefix) ? cookiePrefix : [cookiePrefix];
	for (const prefix of prefixes) {
		const candidates = [`${SECURE_COOKIE_PREFIX}${prefix}.oauth_state`, `${prefix}.oauth_state`];
		for (const name of candidates) {
			const value = parsed?.[name]?.value;
			if (value) return value;
		}
	}
	return null;
}
function getOrigin(scheme) {
	return Linking.createURL("", { scheme });
}
/**
* Compare if session cookies have actually changed by comparing their values.
* Ignores expiry timestamps that naturally change on each request.
*
* @param prevCookie - Previous cookie JSON string
* @param newCookie - New cookie JSON string
* @returns true if session cookies have changed, false otherwise
*/
function hasSessionCookieChanged(prevCookie, newCookie) {
	if (!prevCookie) return true;
	try {
		const prev = JSON.parse(prevCookie);
		const next = JSON.parse(newCookie);
		const sessionKeys = /* @__PURE__ */ new Set();
		Object.keys(prev).forEach((key) => {
			if (key.includes("session_token") || key.includes("session_data")) sessionKeys.add(key);
		});
		Object.keys(next).forEach((key) => {
			if (key.includes("session_token") || key.includes("session_data")) sessionKeys.add(key);
		});
		for (const key of sessionKeys) if (prev[key]?.value !== next[key]?.value) return true;
		return false;
	} catch {
		return true;
	}
}
/**
* Check if the Set-Cookie header contains better-auth cookies.
* This prevents infinite refetching when non-better-auth cookies (like third-party cookies) change.
*
* Supports multiple cookie naming patterns:
* - Default: "better-auth.session_token", "better-auth-passkey", "__Secure-better-auth.session_token"
* - Custom prefix: "myapp.session_token", "myapp-passkey", "__Secure-myapp.session_token"
* - Custom full names: "my_custom_session_token", "custom_session_data"
* - No prefix (cookiePrefix=""): matches any cookie with known suffixes
* - Multiple prefixes: ["better-auth", "my-app"] matches cookies starting with any of the prefixes
*
* @param setCookieHeader - The Set-Cookie header value
* @param cookiePrefix - The cookie prefix(es) to check for. Can be a string, array of strings, or empty string.
* @returns true if the header contains better-auth cookies, false otherwise
*/
function hasBetterAuthCookies(setCookieHeader, cookiePrefix) {
	const cookies = parseSetCookieHeader$1(setCookieHeader);
	const cookieSuffixes = ["session_token", "session_data"];
	const prefixes = Array.isArray(cookiePrefix) ? cookiePrefix : [cookiePrefix];
	for (const name of cookies.keys()) {
		const nameWithoutSecure = stripSecureCookiePrefix(name);
		for (const prefix of prefixes) if (prefix) {
			if (nameWithoutSecure.startsWith(prefix)) return true;
		} else for (const suffix of cookieSuffixes) if (nameWithoutSecure.endsWith(suffix)) return true;
	}
	return false;
}
/**
* Expo secure store does not support colons in the keys.
* This function replaces colons with underscores.
*
* @see https://github.com/better-auth/better-auth/issues/5426
*
* @param name cookie name to be saved in the storage
* @returns normalized cookie name
*/
function normalizeCookieName(name) {
	return name.replace(/:/g, "_");
}
/**
* Max characters written per `setItem`. Native secure stores silently reject
* oversized writes (iOS Keychain refuses values above ~2KB), losing the cookie,
* so a larger value is split across keys here. Mirrors the server's
* `chunkCookie`/`joinChunks` in `session-store.ts`; keep the two in sync.
*
* @see https://github.com/better-auth/better-auth/issues/9151
*/
const STORAGE_VALUE_LIMIT = 1800;
/**
* Marks a base key whose value is split across `<key>.0..N` chunks. The leading
* control char can't start a JSON value (so it never collides) and, unlike NUL,
* survives the native storage bridge without C-string truncation.
*/
const CHUNK_MARKER = "ba-chunks:";
function storageAdapter(storage) {
	return {
		getItem: (name) => {
			const key = normalizeCookieName(name);
			const stored = storage.getItem(key);
			if (stored == null || !stored.startsWith(CHUNK_MARKER)) return stored;
			const count = Number(stored.slice(11));
			if (!Number.isInteger(count) || count < 1) return null;
			let value = "";
			for (let i = 0; i < count; i++) {
				const chunk = storage.getItem(`${key}.${i}`);
				if (chunk == null) return null;
				value += chunk;
			}
			return value;
		},
		setItem: async (name, value) => {
			const key = normalizeCookieName(name);
			try {
				if (value.length <= STORAGE_VALUE_LIMIT) {
					await storage.setItem(key, value);
					return;
				}
				await storage.setItem(key, "");
				const count = Math.ceil(value.length / STORAGE_VALUE_LIMIT);
				for (let i = 0; i < count; i++) {
					const start = i * STORAGE_VALUE_LIMIT;
					await storage.setItem(`${key}.${i}`, value.slice(start, start + STORAGE_VALUE_LIMIT));
				}
				await storage.setItem(key, `${CHUNK_MARKER}${count}`);
			} catch (error) {
				console.error(`[better-auth/expo] failed to persist "${key}" to storage`, error);
			}
		}
	};
}
const expoClient = (opts) => {
	let store = null;
	const storagePrefix = opts?.storagePrefix || "better-auth";
	const cookieName = `${storagePrefix}_cookie`;
	const localCacheName = `${storagePrefix}_session_data`;
	const storage = storageAdapter(opts?.storage);
	const isWeb = Platform.OS === "web";
	const cookiePrefix = opts?.cookiePrefix || "better-auth";
	const clearSessionCache = async () => {
		await storage.setItem(cookieName, "{}");
		store?.atoms.session?.set({
			...store.atoms.session.get(),
			data: null,
			error: null,
			isPending: false
		});
		await storage.setItem(localCacheName, "{}");
	};
	const rawScheme = opts?.scheme || Constants.expoConfig?.scheme || Constants.platform?.scheme;
	const scheme = Array.isArray(rawScheme) ? rawScheme[0] : rawScheme;
	if (!scheme && !isWeb) throw new Error("Scheme not found in app.json. Please provide a scheme in the options.");
	return {
		id: "expo",
		version: PACKAGE_VERSION,
		getActions(_, $store) {
			store = $store;
			const sessionAtom = $store.atoms.session;
			if (!isWeb && !opts?.disableCache && sessionAtom) {
				const raw = storage.getItem(localCacheName);
				const cached = raw ? safeJSONParse(raw) : null;
				const exp = cached?.session?.expiresAt;
				const expMs = exp ? new Date(exp).getTime() : NaN;
				if (!!cached?.user?.id && !!cached.session?.id && expMs > Date.now()) sessionAtom.set({
					...sessionAtom.get(),
					data: cached,
					error: null
				});
			}
			return { getCookie: () => {
				return getCookie(storage.getItem(cookieName) || "{}");
			} };
		},
		fetchPlugins: [{
			id: "expo",
			name: "Expo",
			hooks: { async onSuccess(context) {
				if (isWeb) return;
				const setCookie = context.response.headers.get("set-cookie");
				if (setCookie) {
					if (hasBetterAuthCookies(setCookie, cookiePrefix)) {
						const prevCookie = storage.getItem(cookieName);
						const toSetCookie = getSetCookie(setCookie || "", prevCookie ?? void 0);
						if (hasSessionCookieChanged(prevCookie, toSetCookie)) {
							await storage.setItem(cookieName, toSetCookie);
							store?.notify("$sessionSignal");
						} else await storage.setItem(cookieName, toSetCookie);
					}
				}
				if (context.request.url.toString().includes("/get-session") && !opts?.disableCache) {
					const data = context.data;
					await storage.setItem(localCacheName, JSON.stringify(data));
				}
				if (context.request.url.toString().includes("/sign-out")) await clearSessionCache();
				if (context.data?.redirect && (context.request.url.toString().includes("/sign-in") || context.request.url.toString().includes("/link-social")) && !context.request?.body.includes("idToken")) {
					const to = JSON.parse(context.request.body)?.callbackURL;
					const signInURL = context.data?.url;
					let Browser = void 0;
					try {
						Browser = await import("expo-web-browser");
					} catch {
						try {
							Browser = __require("expo-web-browser");
						} catch (error) {
							throw new Error("\"expo-web-browser\" is not installed as a dependency!", { cause: error });
						}
					}
					if (Platform.OS === "android") try {
						Browser.dismissAuthSession();
					} catch {}
					const oauthStateValue = getOAuthStateValue(storage.getItem(cookieName), cookiePrefix);
					const params = new URLSearchParams({ authorizationURL: signInURL });
					if (oauthStateValue) params.append("oauthState", oauthStateValue);
					const proxyURL = `${context.request.baseURL}/expo-authorization-proxy?${params.toString()}`;
					const result = await Browser.openAuthSessionAsync(proxyURL, to, opts?.webBrowserOptions);
					if (result.type !== "success") return;
					const cookie = new URL(result.url).searchParams.get("cookie");
					if (!cookie) return;
					const toSetCookie = getSetCookie(cookie, storage.getItem(cookieName) ?? void 0);
					await storage.setItem(cookieName, toSetCookie);
					store?.notify("$sessionSignal");
				}
			} },
			async init(url, options) {
				if (isWeb) return {
					url,
					options
				};
				options = options || {};
				options.credentials = "omit";
				if (options.body?.idToken !== void 0) {
					const cookie = url.includes("/link-social") ? getCookie(storage.getItem(cookieName) || "{}") : "";
					options.headers = {
						...options.headers,
						...cookie ? { cookie } : {},
						"x-skip-oauth-proxy": "true"
					};
				} else {
					const cookie = getCookie(storage.getItem(cookieName) || "{}");
					options.headers = {
						...options.headers,
						...cookie ? { cookie } : {},
						"expo-origin": getOrigin(scheme),
						"x-skip-oauth-proxy": "true"
					};
					if (options.body?.callbackURL) {
						if (options.body.callbackURL.startsWith("/")) {
							const url = Linking.createURL(options.body.callbackURL);
							options.body.callbackURL = url;
						}
					}
					if (options.body?.newUserCallbackURL) {
						if (options.body.newUserCallbackURL.startsWith("/")) {
							const url = Linking.createURL(options.body.newUserCallbackURL);
							options.body.newUserCallbackURL = url;
						}
					}
					if (options.body?.errorCallbackURL) {
						if (options.body.errorCallbackURL.startsWith("/")) {
							const url = Linking.createURL(options.body.errorCallbackURL);
							options.body.errorCallbackURL = url;
						}
					}
					if (url.includes("/sign-out")) await clearSessionCache();
				}
				return {
					url,
					options
				};
			}
		}]
	};
};
//#endregion
export { expoClient, getCookie, getSetCookie, hasBetterAuthCookies, normalizeCookieName, parseSetCookieHeader, setupExpoFocusManager, setupExpoOnlineManager, storageAdapter };
